### PR TITLE
Player physics: Ensure larger dtime simulation steps

### LINKED ISSUE #10543: get_player_velocity() inconsistently returns non-null vector for static player
##### Summary

`player:get_player_velocity()` returns non-null vector even if the player is not moving at all and touching ground (negative Y-component), see [this video](https://youtu.be/AGh4-vq5_Xc) - other times, it returns a constant null vector.

##### Steps to reproduce

```lua
minetest.register_globalstep(function()
    for _, player in pairs(minetest.get_connected_players()) do
        minetest.chat_send_all(minetest.pos_to_string(player:get_player_velocity()))
    end
end)
```

### PR DESCRIPTION
**Pull request by appgurueu**
Should fix #10543 by fixing [this underlying issue](https://github.com/minetest/minetest/issues/10543#issuecomment-716711613) mentioned by @TheTermos

### CODE REVIEW

**Comment by lhofhansl** [comment]
Hi @appgurueu can you briefly explain how this fixes the problem mentioned?

**Comment by appgurueu** [comment]
Pretty simple. It used to work by dividing the dtime into smaller dtimes, then simulating. This is pretty standard. But, movement code apparently is susceptible to very small leftover dtimes, which this occasionally produces.

This fix now does not generate small leftover dtimes anymore: it guarantees that any leftover dtime is > dtime_max_increment/2

**Comment by TheTermos** [comment]
Can you confirm this solves the issue?

Can it be rephrased so it's more logical and less mysterious, maybe something like:
```
    int slices;
    slices = int(dtime/dtime_max_increment) + (std::fmod(dtime,dtime_max_increment)>0)*1;
    f32 dtime_part=dtime/slices;
    
    for(int i=0;i<slices;i++){
     ...dotheloopstuff...
    }
```
 this way it's immediately apparent what it does, I think.

**Comment by appgurueu** [comment]
> Can you confirm this solves the issue?

Test for yourself, but I have not seen the issue again.

> Can it be rephrased so it's more logical and less mysterious

Yes

**Comment by lhofhansl** [comment]
Thanks for explaining @appgurueu .

File: src/client/clientenvironment.cpp [hunk 1]
```diff
@@ -171,50 +171,28 @@ void ClientEnvironment::step(float dtime)
 
 	// Maximum time increment (for collision detection etc)
 	// time = distance / speed
-	f32 dtime_max_increment = 1;
+	f32 dtime_min_step = 1;
 	if(player_speed > 0.001)
-		dtime_max_increment = position_max_increment / player_speed;
+		dtime_min_step = position_max_increment / player_speed;
 
 	// Maximum time increment is 10ms or lower
```
Comment (sfan5) [inline src/client/clientenvironment.cpp:178]: comment is not longer accurate

File: src/client/clientenvironment.cpp [hunk 2]
```diff
@@ -171,50 +171,28 @@ void ClientEnvironment::step(float dtime)
 
 	// Maximum time increment (for collision detection etc)
```
Comment (sfan5) [inline src/client/clientenvironment.cpp:172]: same here

**Comment by lhofhansl** [comment]
The new code is certainly "less mysterious", so a good change.
Looks good to - except for comments that @sfan5 mentioned.

**Comment by appgurueu** [comment]
Renamed the variable back to `dtime_max_increment`.

**Review by sfan5 — APPROVED** [review]

**Comment by lhofhansl** [comment]
This looks good to me.
Could you rebase?

**Comment by sfan5** [comment]
It can be merged correctly if squashing, `git am` unfortunately doesn't work.

**Comment by paramat** [comment]
I removed the 'testing lneeded' label i added, as i cannot see how this particularly needs testing, seeing as `dtime_part` now has a smaller range of values than before.
I like the new implementation of this.

**Pull request merged**